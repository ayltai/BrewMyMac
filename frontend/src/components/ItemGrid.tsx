import { Grid, } from '@mui/material';
import React from 'react';

import type { Item, } from '../models';

import { SelectableItem, } from './SelectableItem';

export const ItemGrid = ({
    items,
} : {
    items : Item[],
}) => (
    <Grid
        container
        padding={2}
        spacing={2}>
        {items.map((item : Item, index : number) => (
            <Grid
                key={`${item.id}-${item.source}-${index}`}
                item
                xs={6}
                sm={4}
                lg={3}>
                <SelectableItem item={item} />
            </Grid>
        ))}
    </Grid>
);
