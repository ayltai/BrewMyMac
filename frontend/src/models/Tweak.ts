export type Tweak = {
    id            : string,
    name          : string,
    description?  : string,
    author?       : string,
    defaultValue? : string | number | boolean,
};
