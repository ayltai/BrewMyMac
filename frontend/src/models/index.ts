export type { AppStoreSearchResult, AppStoreSearchResults, } from './AppStoreSearchResults';
export type { HomebrewFormula, } from './HomebrewFormulae';
export type { Item, } from './Item';
export type { Tweak, } from './Tweak';
