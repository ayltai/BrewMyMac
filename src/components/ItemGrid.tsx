import { Grid, } from '@mui/material';
import React from 'react';

import type { Item, } from '../types';
import { default as SelectableItem, } from './SelectableItem';

export const ItemGrid = ({
    items,
} : {
    items : Item[],
}) => (
    <Grid
        container
        padding={2}
        spacing={2}>
        {items.map(item => (
            <Grid
                item
                key={`item-${item.source}-${item.id}`}
                xs={12}
                sm={6}
                md={4}
                lg={3}>
                <SelectableItem item={item} />
            </Grid>
        ))}
    </Grid>
);
