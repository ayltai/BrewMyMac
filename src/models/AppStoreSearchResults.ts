export type AppStoreSearchResult = {
    artworkUrl60?  : string,
    artworkUrl100? : string,
    artworkUrl512? : string,
    artistName?    : string,
    trackId        : string,
    trackName      : string,
    description?   : string,
};

export type AppStoreSearchResults = {
    resultCount : number,
    results     : AppStoreSearchResult[],
};
