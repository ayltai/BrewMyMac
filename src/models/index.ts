export type { AppStoreSearchResult, AppStoreSearchResults, } from './AppStoreSearchResults';
export type { HomebrewFormula, } from './HomebrewFormulae';
export type { Item, } from './Item';
export type { Session, } from './Session';
export type { Tweak, } from './Tweak';
