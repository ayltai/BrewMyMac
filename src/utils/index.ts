export { createEmotionCache, } from './createEmotionCache';
export { handleError, } from './handleError';
