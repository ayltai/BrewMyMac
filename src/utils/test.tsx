import { createTheme, ThemeProvider, } from '@mui/material';
import { configureStore, PreloadedState, Store, } from '@reduxjs/toolkit';
import '@testing-library/jest-dom/extend-expect';
import { act, render, } from '@testing-library/react';
import React, { ReactNode, } from 'react';
import { Provider, } from 'react-redux';

import { appStoreApi, homebrewApi, tweakApi, } from '../apis';
import type { AppStoreSearchResults, HomebrewFormula, Tweak, } from '../models';
import { sessionReducer, store as reduxStore, } from '../redux';
import { palette, } from '../styles';

const INITIAL_FILTER = 'firefox';

export const mockApis = async () => {
    const HOMEBREW_API_RESPONSE : HomebrewFormula[] = [
        {
            token     : `${INITIAL_FILTER} homebrew token`,
            name      : [
                `${INITIAL_FILTER} homebrew name`,
            ],
            full_name : `${INITIAL_FILTER} homebrew full name`,
        },
    ];

    const HOMEBREW_CAST_API_RESPONSE : HomebrewFormula[] = [
        {
            token     : `${INITIAL_FILTER} homebrew cask token`,
            name      : [
                `${INITIAL_FILTER} homebrew cask name`,
            ],
            full_name : `${INITIAL_FILTER} homebrew cask full name`,
        },
    ];

    const TWEAKS_API_RESPONSE : Tweak[] = [
        {
            id   : '2',
            name : `${INITIAL_FILTER} tweak name`,
        },
    ];

    const APP_STORE_API_RESPONSE : AppStoreSearchResults = {
        resultCount : 1,
        results     : [
            {
                trackId   : '1',
                trackName : `${INITIAL_FILTER} app store name`,
            },
        ],
    };

    fetchMock.doMock(request => {
        if (request.url.startsWith('https://formulae.brew.sh/api/formula.json')) {
            return Promise.resolve({
                status : 200,
                body   : JSON.stringify(HOMEBREW_API_RESPONSE),
            });
        }

        if (request.url.startsWith('https://formulae.brew.sh/api/cask.json')) {
            return Promise.resolve({
                status : 200,
                body   : JSON.stringify(HOMEBREW_CAST_API_RESPONSE),
            });
        }

        if (request.url.startsWith('https://raw.githubusercontent.com')) {
            return Promise.resolve({
                status : 200,
                body   : JSON.stringify(TWEAKS_API_RESPONSE),
            });
        }

        if (request.url.startsWith('https://search-itunes.vercel.app')) {
            return Promise.resolve({
                status : 200,
                body   : JSON.stringify(APP_STORE_API_RESPONSE),
            });
        }

        return Promise.resolve({
            status : 404,
        });
    });

    await act(async () => {
        await reduxStore.dispatch(homebrewApi.endpoints.formula.initiate());
        await reduxStore.dispatch(homebrewApi.endpoints.cask.initiate());
        await reduxStore.dispatch(tweakApi.endpoints.getTweaks.initiate());
        await reduxStore.dispatch(appStoreApi.endpoints.search.initiate(`${INITIAL_FILTER} app store name`));
    });
};

const createStore = (preloadedState? : Record<string, any>) => configureStore({
    reducer : {
        session : sessionReducer,
    },
    preloadedState,
});

const defaultStore = createStore();

type RootState = ReturnType<typeof defaultStore.getState>;

const customRender = (ui : any, {
    preloadedState,
    store = createStore(preloadedState),
    ...rest
} : {
    preloadedState?   : PreloadedState<RootState>,
    store?            : Store<RootState>,
    [ rest : string ] : any,
} = {}) => render(ui, {
    wrapper : ({
        children,
    } : {
        children : ReactNode,
    }) => {
        const theme = createTheme({
            palette : {
                ...palette,
                mode : 'light',
            },
        });

        return (
            <Provider store={store}>
                <ThemeProvider theme={theme}>
                    {children}
                </ThemeProvider>
            </Provider>
        );
    },
    ...rest,
});

export * from '@testing-library/react';

export { customRender as render, };
