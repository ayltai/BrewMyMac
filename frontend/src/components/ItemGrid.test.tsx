import React from 'react';

import { Item, } from '../models';
import { render, } from '../utils/test';

import { ItemGrid, } from './ItemGrid';

const ITEM : Item = {
    id        : 'dummy_id',
    name      : 'dummy_name',
    source    : 'Tweak',
    parameter : 'dummy_parameter',
};

describe('<ItemGrid />', () => {
    it('renders correctly', () => expect(render(<ItemGrid items={[
        ITEM,
    ]} />).asFragment()).toMatchSnapshot());
});
