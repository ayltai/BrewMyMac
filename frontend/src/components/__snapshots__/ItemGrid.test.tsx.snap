// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ItemGrid /> renders correctly 1`] = `
<DocumentFragment>
  <div
    class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-2 css-1ebqi03-MuiGrid-root"
  >
    <div
      class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-6 MuiGrid-grid-sm-4 MuiGrid-grid-lg-3 css-16cooyn-MuiGrid-root"
    >
      <div
        class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-7f9rp6-MuiPaper-root-MuiCard-root"
      >
        <button
          class="MuiButtonBase-root MuiCardActionArea-root css-b6lsxj-MuiButtonBase-root-MuiCardActionArea-root"
          role="gridcell"
          tabindex="0"
          type="button"
        >
          <div
            class="MuiCardContent-root css-46bh2p-MuiCardContent-root"
          >
            <h6
              class="MuiTypography-root MuiTypography-h6 css-1lt47b1-MuiTypography-root"
            >
              dummy_name
            </h6>
            <div
              class="MuiBox-root css-15p9n5u"
            />
            <p
              class="MuiTypography-root MuiTypography-body2 css-l1juks-MuiTypography-root"
            />
            <div
              class="MuiBox-root css-164r41r"
            >
              <div
                class="MuiChip-root MuiChip-filled MuiChip-sizeSmall MuiChip-colorDefault MuiChip-filledDefault css-10rs1bw-MuiChip-root"
              >
                <span
                  class="MuiChip-label MuiChip-labelSmall css-wjsjww-MuiChip-label"
                >
                  Tweak
                </span>
              </div>
            </div>
          </div>
          <span
            class="MuiCardActionArea-focusHighlight css-1v2exvi-MuiCardActionArea-focusHighlight"
          />
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
        <div
          class="MuiCardActions-root MuiCardActions-spacing css-1t6e9jv-MuiCardActions-root"
        >
          <button
            class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeSmall MuiButton-textSizeSmall css-fy9mre-MuiButtonBase-root-MuiButton-root"
            tabindex="0"
            type="button"
          >
            action_customize
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
